src/main/scala/com/gradle/internal/CustomBuildScanEnhancements.scala

```diff
@@ -0,0 +1,341 @@
+        if (ideaVendorNameValue == "JetBrains") tagIde("IntelliJ IDEA", getOrEmpty(sysProperty(SYSTEM_PROP_IDEA_VERSION)))
+      }
+      else if (sysProperty(SYSTEM_PROP_IDEA_VERSION).isDefined) {
+        // this case should be handled by the ideaVendorName condition but keeping it for compatibility reason (ideaVendorName started with 2020.1)
+        tagIde("IntelliJ IDEA", sysProperty(SYSTEM_PROP_IDEA_VERSION).get)
+      }
+      else if (sysProperty(SYSTEM_PROP_IDEA_MANAGED).isDefined) {
+        tagIde("IntelliJ IDEA", "")
+      }
+      else if (sysProperty(SYSTEM_PROP_ECLIPSE_BUILD_ID).isDefined) {
+          tagIde("Eclipse", sysProperty(SYSTEM_PROP_ECLIPSE_BUILD_ID).get)
+      }
+      else buildScan.tag("Cmd Line")
+
+      ifDefined(sysProperty(SYSTEM_PROP_IDEA_SYNC_ACTIVE), (_: String) => buildScan.tag("IDE sync"))
+    }
+  }
+
+  private def captureCiOrLocal(): Unit = {
+    buildScan.tag(if (isCi) "CI" else "LOCAL")
+  }
+
+  private def captureCiMetadata(): Unit = {
+      if (isJenkins || isHudson) {
+          val buildUrl: Option[String] = envVariable("BUILD_URL")
+          val buildNumber: Option[String] = envVariable("BUILD_NUMBER")
+          val nodeName: Option[String] = envVariable("NODE_NAME")
+          val jobName: Option[String] = envVariable("JOB_NAME")
+          val stageName: Option[String] = envVariable("STAGE_NAME")
+          if (buildUrl.isDefined) {
+              buildScan.link(if (isJenkins) {
+                  "Jenkins build"
+              }
+              else {
+                  "Hudson build"
+              }, buildUrl.get)
+          }
+          ifDefined(buildNumber, (value: String) => buildScan.addValue("CI build number", value))
+          ifDefined(nodeName, (value: String) => addCustomValueAndSearchLink("CI node", value))
+          ifDefined(jobName, (value: String) => addCustomValueAndSearchLink("CI job", value))
+          ifDefined(stageName, (value: String) => addCustomValueAndSearchLink("CI stage", value))
+          ifDefined(jobName, (j: String) => ifDefined(buildNumber, (b: String) => {
+              val params: mutable.Map[String, String] = mutable.Map.empty
+              params += "CI job" -> j
+              params += "CI build number" -> b
+              addSearchLink("CI pipeline", params.toMap)
+          }))
+      }
+
+      if (isTeamCity) {
+          val teamcityBuildPropertiesFile = envVariable("TEAMCITY_BUILD_PROPERTIES_FILE")
+          if (teamcityBuildPropertiesFile.isDefined) {
+              val buildProperties = readPropertiesFile(teamcityBuildPropertiesFile.get)
+              val teamCityBuildId = Option.apply(buildProperties.getProperty("teamcity.build.id"))
+              if (isNotEmpty(teamCityBuildId)) {
+                  val teamcityConfigFile = Option.apply(buildProperties.getProperty("teamcity.configuration.properties.file"))
+                  if (isNotEmpty(teamcityConfigFile)) {
```

Comment:
Hm, while I agree it would be nice to simplify it, there's also a notion that the CCUD implementation should be as similar as possible..while at the same time this is written in scala and the other two are in java... What do you think Nelson?

Reply:
This is a good question. We had similar discussions before when trying to align groovy and kotlin buildscripts. We should use idiomatic Scala whenever possible because scala customers will be expecting the idiomatic version and this shows customers that we experts in that domain.